flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/parser/SqlClientSyntaxHighlighter.java

```diff
@@ -0,0 +1,256 @@
+                "`",
+                (dialect) -> dialect == SqlDialect.DEFAULT || dialect == null,
+                (asb, style) -> asb.style(style.getSqlIdentifierStyle())),
+        HIVE_SQL_QUOTED_IDENTIFIER(
+                2,
+                "\"",
+                "\"",
+                (dialect) -> dialect == SqlDialect.HIVE,
+                (asb, style) -> asb.style(style.getSqlIdentifierStyle())),
+        ONE_LINE_COMMENTED(
+                3, "--", "\n", dialect -> true, (asb, style) -> asb.style(style.getCommentStyle())),
+        HINTED(4, "/*+", "*/", dialect -> true, (asb, style) -> asb.style(style.getHintStyle())),
+        MULTILINE_COMMENTED(
+                5, "/*", "*/", dialect -> true, (asb, style) -> asb.style(style.getCommentStyle())),
+        DEFAULT(
+                Integer.MAX_VALUE,
+                null,
+                null,
+                dialect -> true,
+                (asb, style) -> asb.style(style.getDefaultStyle()));
+
+        private final String start;
+        private final String end;
+        private final Function<SqlDialect, Boolean> condition;
+
+        private final int order;
+
+        private final BiConsumer<AttributedStringBuilder, SyntaxHighlightStyle> styleSetter;
+
+        private static final List<State> STATE_LIST_WITHOUT_DEFAULT =
+                Arrays.stream(State.values())
+                        .filter(t -> t != DEFAULT)
+                        .sorted(Comparator.comparingInt(o -> o.order))
+                        .collect(Collectors.toList());
+        private static final Set<Character> STATE_START_SYMBOLS =
+                Arrays.stream(State.values())
+                        .filter(t -> t != DEFAULT)
+                        .map(t -> t.start.charAt(0))
+                        .collect(Collectors.toSet());
+
+        State(
+                int order,
+                String start,
+                String end,
+                Function<SqlDialect, Boolean> condition,
+                BiConsumer<AttributedStringBuilder, SyntaxHighlightStyle> styleSetter) {
+            this.start = start;
+            this.end = end;
+            this.order = order;
+            this.condition = condition;
+            this.styleSetter = styleSetter;
+        }
+
+        static State computeStateAt(String input, int pos, SqlDialect dialect) {
+            final char currentChar = input.charAt(pos);
+            if (!STATE_START_SYMBOLS.contains(currentChar)) {
+                return DEFAULT;
+            }
+            for (State state : STATE_LIST_WITHOUT_DEFAULT) {
+                if (state.condition.apply(dialect)
+                        && state.start.regionMatches(0, input, pos, state.start.length())) {
+                    return state;
+                }
+            }
+            return DEFAULT;
+        }
+
+        boolean isEndOfState(String input, int pos) {
+            if (this == DEFAULT) {
+                return false;
+            }
```

Comment:
nit: BTW, it's not correct here because the other state's beginning is the end of the current state. Or do we just throw an exception to notify others not to use this for the `DEFAULT` state?